File: src/main/kotlin/com/xiaohunao/minecraftdevelopmenttoolkit/ws/MessageDispatcher.kt
package com.xiaohunao.minecraftdevelopmenttoolkit.ws

import com.intellij.openapi.diagnostic.Logger
import com.xiaohunao.mdt.protocol.Message
import com.xiaohunao.mdt.protocol.MessageType

class MessageDispatcher {

    private val logger = Logger.getInstance(MessageDispatcher::class.java)

    private val helloHandlers = mutableListOf<(Message) -> Unit>()
    private val uiRenderHandlers = mutableListOf<(String, Message) -> Unit>()
    private val stateHandlers = mutableListOf<(String, Message) -> Unit>()
    private val stateSubscribeHandlers = mutableListOf<(String, Message) -> Unit>()
    private val tabHandlers = mutableListOf<(Message) -> Unit>()
    private val notificationHandlers = mutableListOf<(Message) -> Unit>()
    private val consoleHandlers = mutableListOf<(String, Message) -> Unit>()
    private val commandSuggestHandlers = mutableListOf<(Message) -> Unit>()

    fun onHello(handler: (Message) -> Unit) { helloHandlers.add(handler) }
    fun onUiRender(handler: (tabId: String, message: Message) -> Unit) { uiRenderHandlers.add(handler) }
    fun onStateUpdate(handler: (tabId: String, message: Message) -> Unit) { stateHandlers.add(handler) }
    fun onStateSubscribe(handler: (tabId: String, message: Message) -> Unit) { stateSubscribeHandlers.add(handler) }
    fun onTabChange(handler: (message: Message) -> Unit) { tabHandlers.add(handler) }
    fun onNotification(handler: (message: Message) -> Unit) { notificationHandlers.add(handler) }
    fun onConsoleAppend(handler: (tabId: String, message: Message) -> Unit) { consoleHandlers.add(handler) }
    fun onCommandSuggestResponse(handler: (Message) -> Unit) { commandSuggestHandlers.add(handler) }

    fun dispatch(message: Message) {
        val tabId = message.tabId

        when (message.type) {
            MessageType.HELLO -> {
                helloHandlers.forEach { it(message) }
            }
            MessageType.UI_RENDER -> {
                if (tabId != null) uiRenderHandlers.forEach { it(tabId, message) }
            }
            MessageType.STATE_SNAPSHOT, MessageType.STATE_DELTA -> {
                if (tabId != null) stateHandlers.forEach { it(tabId, message) }
            }
            MessageType.TAB_OPEN, MessageType.TAB_CLOSE, MessageType.TAB_UPDATE -> {
                tabHandlers.forEach { it(message) }
            }
            MessageType.NOTIFICATION -> {
                notificationHandlers.forEach { it(message) }
            }
            MessageType.CONSOLE_APPEND -> {
                if (tabId != null) consoleHandlers.forEach { it(tabId, message) }
            }
            MessageType.EVENT_ACK -> {
                logger.info("Event ACK received: ${message.requestId}")
            }
            MessageType.STATE_SUBSCRIBE -> {
                // CLIENT_TO_SERVER — plugin normally only sends these, never receives them.
                // Handle gracefully in case one arrives unexpectedly.
                if (tabId != null) stateSubscribeHandlers.forEach { it(tabId, message) }
            }
            MessageType.COMMAND_SUGGEST_RESPONSE -> {
                commandSuggestHandlers.forEach { it(message) }
            }
            MessageType.ERROR -> {
                logger.error("Server error: ${message.payload}")
            }
            else -> {
                logger.warn("Unhandled message type: ${message.type}")
            }
        }
    }
}

File: web/src/toolwindow/MessageDispatcher.ts
import type { Message } from '../types/protocol'

type TabMessageHandler = (tabId: string, message: Message) => void
type GlobalMessageHandler = (message: Message) => void

export class MessageDispatcher {
  private helloHandlers: GlobalMessageHandler[] = []
  private uiRenderHandlers: TabMessageHandler[] = []
  private stateHandlers: TabMessageHandler[] = []
  private tabHandlers: GlobalMessageHandler[] = []
  private notificationHandlers: GlobalMessageHandler[] = []
  private consoleHandlers: TabMessageHandler[] = []
  private commandSuggestHandlers: GlobalMessageHandler[] = []

  onHello(handler: GlobalMessageHandler) { this.helloHandlers.push(handler) }
  onUiRender(handler: TabMessageHandler) { this.uiRenderHandlers.push(handler) }
  onStateUpdate(handler: TabMessageHandler) { this.stateHandlers.push(handler) }
  onTabChange(handler: GlobalMessageHandler) { this.tabHandlers.push(handler) }
  onNotification(handler: GlobalMessageHandler) { this.notificationHandlers.push(handler) }
  onConsoleAppend(handler: TabMessageHandler) { this.consoleHandlers.push(handler) }
  onCommandSuggestResponse(handler: GlobalMessageHandler) { this.commandSuggestHandlers.push(handler) }

  dispatch(message: Message) {
    const tabId = message.tabId

    switch (message.type) {
      case 'HELLO':
        this.helloHandlers.forEach(h => h(message))
        break
      case 'UI_RENDER':
        if (tabId) this.uiRenderHandlers.forEach(h => h(tabId, message))
        break
      case 'STATE_SNAPSHOT':
      case 'STATE_DELTA':
        if (tabId) this.stateHandlers.forEach(h => h(tabId, message))
        break
      case 'TAB_OPEN':
      case 'TAB_CLOSE':
      case 'TAB_UPDATE':
        this.tabHandlers.forEach(h => h(message))
        break
      case 'NOTIFICATION':
        this.notificationHandlers.forEach(h => h(message))
        break
      case 'CONSOLE_APPEND':
        if (tabId) this.consoleHandlers.forEach(h => h(tabId, message))
        break
      case 'COMMAND_SUGGEST_RESPONSE':
        this.commandSuggestHandlers.forEach(h => h(message))
        break
      case 'EVENT_ACK':
        break
      case 'ERROR':
        console.error('[MDT] Server error:', message.payload)
        break
      default:
        console.warn('[MDT] Unhandled message type:', message.type)
    }
  }
}
